src/cart/crt_swim.c

```diff
@@ -26,10 +26,22 @@
 #define D_LOGFAC	DD_FAC(swim)
 #define CRT_USE_GURT_FAC
 
+#include <ctype.h>
 #include "crt_internal.h"
 #include "swim/swim_internal.h"
 
 #define CRT_OPC_SWIM_VERSION	0
+#define CRT_SWIM_FAIL_BASE	((CRT_OPC_SWIM_PROTO >> 16) | \
+				 (CRT_OPC_SWIM_VERSION << 4))
+#define CRT_SWIM_FAIL_DROP_RPC	(CRT_SWIM_FAIL_BASE | 0x1)
+
+/**
+ * use this macro to determine if a fault should be injected at
+ * a specific place
+ */
+#define CRT_SWIM_SHOULD_FAIL(fa, id)				\
```

Comment:
This should probably call D_SHOULD_FAIL() first, then do the other checks.  Is "id" a rank, it's not clear from the names.

D_SHOUL_FAIL is defined to be efficient if not used, and to do logging if it is so it's useful to call it first.

Reply:
yes, SWIM id is rank in current implementation.

Reply:
Function call is more expensive and I'd like to make it last.

Reply:
D_SHOULD_FAIL() is not a function call, as I said it's deliberately written to be very low-cost when FI is disabled.

Reply:
Actually my macro join D_SHOULD_FAIL with my addition check. 
D_SHOULD_FAIL() is just line 44 ->  "d_fault_inject && d_should_fail(fa)"

Reply:
I add conditional SWIM specific. So, I cannot change D_ macro.

Reply:
What I mean here is why can't you do something like this instead?
crt_swim_should_fail && (crt_swim_fail_id == id) && D_SHOULD_FAIL(fa) ?

Reply:
The D_SHOULD_FAIL(fa) is not just conditions. It also has a THEN block. So, I don't want to have additional overhead on every  check.

Reply:
The code block in D_SHOULD_FAIL is only executed if fault injection is turned on, and does logging which is something that we do want to happen, and in a consistent way.